셀 컴포넌트 렌더링 구조 변경 및 유틸 디렉토리 생성

ImuruKevol opened the pull request:
셀 컴포넌트 렌더링시 콜백으로 렌더링 되던 구조에서 콜백 구조 삭제
- 콜백 구조 삭제에 따른 inputRef 적용 메커니즘 변경
  - 스토어에 저장 후 조건에 맞는 셀에서만 불러와 적용
- cellIndex를 uuid로 변경함에 따라 uuid와 index를 매핑하는 매니저 클래스 생성
  - 스토어에 저장하여 계속 관리
기존 리듀서에서 사용되던 splice 함수를 utils로 이동

Commit:
Feat: 셀 클릭시 포커스 인덱스 이동

셀 클릭시 현재 포커스 인덱스 이동 기능 추가

Commit:
Merge: branch 'develop' of https://github.com/connect-foundation/2019-14 into develop

Commit:
Feat: 빈 셀에서 엔터 입력, 셀의 맨 앞에서 백스페이스 입력 기능 추가

- 빈 셀에서 엔터 입력시 기본 입력 셀로 초기화
  - 현재 태그 기능 미구현으로 인해 비활성화 상태
- 셀의 맨 앞에서 백스페이스 입력시 기본 셀로 초기화
  - 기본 셀일 때, 변환된 셀일 때 동작 구현 필요(아직 미구현 상태)

Commit:
Merge: branch 'develop' of https://github.com/connect-foundation/2019-14 into develop

Commit:
Merge: branch 'cell/transform' into develop

Commit:
Refactor: 셀 관련 플럭스 패턴과 마크다운 룰 머지

셀 관련 플럭스 패턴과 마크다운 룰, 디자인 소스 머지

Commit:
Chore: 컴포넌트 매핑 렌더링시 사용할 uuid 모듈 설치

uuidv4 모듈 사용 목적
- 컴포넌트를 맵 함수로 렌더링 시 사용
- 컴포넌트 렌더링시 인덱스 판별시 사용

Commit:
Refactor: 셀 컴포넌트 관련 캐싱 관련 구조 변경

셀 컴포넌트 저장시 콜백을 저장하는 형태에서 컴포넌트를 저장하도록 변경
- 이에 따라 inputRef를 스토어에 저장하여 원하는 타이밍에 가져다 쓰도록
변경
- 기존의 셀 렌더링시 cellIndex를 부여하는 형태에서 콜백 삭제로 인하여
따로 uuid를 생성하고, 그를 관리하는 뼈대 코드 작성(cellReducer.js)
  - 리팩토링시 util에 분리 예정

Commit:
Refactor: 셀 스토어 구조 변경 및 유틸 디렉토리 생성

현재 셀이 인덱스가 무엇인지 알기 위해 cellUuid를 관리하는 매니저 생성
- 스토어에 저장하여 계속 관리
cellReducer.js에 존재하던 splice 함수를 utils의 index.js로 이동
- 간단한 기능들은 index.js에 작성
- Usage
```
import { splice } from "../utils";
```

Commit:
Refactor: utils의 index.js 서식 변경

범용적인 공통 함수를 utils로 사용
특별한 함수 및 클래스는 import하여 export

RBJH commented on client/src/actions/CellAction.js:
```diff
@@ -29,62 +30,85 @@ const cellActionCreator = {
    * - 위 방향키: 이전 셀로 포커스 이동
    */
-  prev() {
+  focusPrev() {
```
함수명 상세히 수정하신 부분 좋네요!:+1:

RBJH commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -38,57 +45,83 @@ const MarkdownWrapper = styled.input`
 
   border-left: ${(props) => {
     if (props.isQuote) return "0.25rem solid silver";
+    return null;
```
이부분 삼항연산자로 수정했습니다!👏

RBJH commented on client/src/components/editor/MarkdownRenderer.jsx:
```diff
@@ -16,80 +13,99 @@ const useCellDispatch = () => {
+  const { currentIndex, uuidManager, cursor } = cellState;
+  let inputRef = null;
+
+  const cellIndex = uuidManager.findIndex(cellUuid);
```
파라미터 명은 cellId면 충분해보입니다.😅

ImuruKevol replied:
의미상은 index가 맞기는 합니다.
uuid로 index를 찾아서 오거든요

RBJH reviewed (commented):
고생하셨습니다!👍👍

DrizzlingCattus commented on client/src/components/editor/MarkdownRenderer.jsx:
```diff
@@ -16,80 +13,99 @@ const useCellDispatch = () => {
+    // const cellIndex = uuidManager.findIndex(cellUuid);
+    // if (currentIndex === cellIndex) {
+    //   inputRef = cellState.inputRef;
+    //   console.log(cellIndex, currentIndex);
```
주석은 지워서 커밋해주시면 감사하겠습니다!

ImuruKevol replied:
이 부분은 주석 삭제 안한게 실수가 맞네요ㅠ

DrizzlingCattus commented on client/src/components/editor/EditorComponent.jsx:
```diff
@@ -10,20 +11,20 @@ const EditorComponent = () => {
   const inputRef = useRef(null);
 
   useEffect(() => {
+    const renderTarget = (uuid) => <MarkdownTransformer cellUuid={uuid} />;
```
콜백으로 넘겨준다면 이름을 그에 맞게 변경해야겠네요.

DrizzlingCattus commented on client/src/components/editor/MarkdownRenderer.jsx:
```diff
@@ -129,30 +169,30 @@ const MarkdownTransformer = ({ inputRef }) => {
       onKeyDown={keyDownHandler}
       onFocus={focusHandler}
-      ref={inputRef}
+      inputRef={inputRef}
```
이거 잘되나요? ref 리엑트 속성이어야 ref로 붙을텐데

ImuruKevol replied:
EditorInput 컴포넌트에 props로 넘겨주는거라 상관은 없습니다.
다만 이제 EditorInput을 한단계 끌어올릴 예정입니다.

dimsssss commented on client/src/components/editor/EditorInput.jsx:
```diff
@@ -38,57 +45,83 @@ const MarkdownWrapper = styled.input`
+  //       />
+  //     </MarkdownWrapper>
+  //   );
+  // }
```
특별한 용도가 없다면 삭제하는게 나을 듯 싶네요

ImuruKevol replied:
원래 돌아가야 하는 코드인데 버그때문에 잠시 주석을 해놓은건데 설명을 달아놓는걸 깜빡했네요

dimsssss reviewed (commented):
특정 파일에 안 쓰는 것처럼 보이는 코드를 주석 처리한 부분이 많았습니다.

고생하셨습니다

Commit:
Fix: PR 코멘트 반영

PR 코멘트 반영
- MarkdownRenderer의 useEffect 내의 주석 삭제
- EditorComponent의 콜백 넣는 변수명 변경
